tuist/tuist
# Ensure static precompiled dependencies are only linked in targets that support linking

Resolves https://github.com/tuist/tuist/issues/4859

### Short description 📝

- Static precompiled targets were currently being "linked" directly by the target that declared a dependency on them
  - note: linking only really took place if the target was one that supports linking (e.g. app, unit tests, etc..), for any static targets the symbols of the precomplied binary symbols would be included
- This worked in simple cases where only one target would depend on the precompiled static target
- In scenarios where multiple static targets needed to depend on the same precompiled static target, build failures would occur due to duplicate symbols
- To solve this, the precompiled static product is now only linked in targets that support linking and instead they are marked as dependencies of the static targets to ensure they can compile
  - For `XCFramework`s those are added to the "Dependencies" copy products build phase
  - For precompiled `Framework`s they are added to the framework search paths _(no change from before the changes)_
  - For precompiled libraries (`.a`) 
     - They are added to the library and swift include search paths of the static targets that depend on them _(no change from before the changes)_
     - Additionally, they are also added to the library search paths of the final linking target

### How to test the changes locally 🧐

XCFramework Test Plan:

- Build this version of tuist

```sh
swift build
```

- Generate the fixture `ios_app_with_xcframeworks`

```sh
swift run tuist generate --path projects/tuist/fixtures/ios_app_with_xcframeworks
```

- Build the project and verify it succeeds
- Verify the precompiled `MyStaticLibrary.xcframework` is only linked in the `App` target
- Verify the precompiled `MyStaticLibrary.xcframework` is declared a dependency (but not in the link phase) of target `StaticFrameworkA`
- Verify `StaticFrameworkA` build successfully in isolation 
- Verify `App` builds and runs successfully without any duplicate symbol errors 

Framework Test Plan:

- Generate the fixture `ios_app_with_static_frameworks`

```sh
swift run tuist generate --path projects/tuist/fixtures/ios_app_with_static_frameworks
```

- Build the project and verify it succeeds
- Verify the precompiled static framework is only linked in the `App` target
- Verify the precompiled static framework is not in the link phase of target `A`
- Verify building the target `A` in isolation still works

### Contributor checklist ✅

- ~In case the PR introduces changes that affect users, the documentation has been updated~

### Reviewer checklist ✅
